### PR TITLE
Added views and controllers for confirm email and verification email sent

### PR DESCRIPTION
**Pull request by davidmthornton**

### CODE REVIEW

**Comment by TomDudley92** [comment]
Forgot to mention in the review. There needs to be some integration tests for the new pages

File: test/controllers/agent/ConfirmEmailControllerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        val result = TestConfirmEmailController.updateEmailAddress()(request)
+
+        status(result) shouldBe Status.SEE_OTHER
+        redirectLocation(result) shouldBe Some("/vat-through-software/representative/client-vat-number")
+
+      }
+    }
+
+    "there is a non-verified email in session" should {
+      "redirect the user to the send email verification pagee" in {
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply test/controllers/agent/ConfirmEmailControllerSpec.scala]: This needs to be addressed

File: app/controllers/agent/ConfirmEmailController.scala [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    extractSessionEmail(agent) match {
+      case Some(email) =>
+        Ok(views.html.agent.confirm_email(email))
+      case _ =>
+        Redirect(routes.CaptureEmailController.show())
+    }
+  }
+
+  def updateEmailAddress(): Action[AnyContent] = authenticate { implicit agent =>
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply app/controllers/agent/ConfirmEmailController.scala]: This needs to be addressed

File: test/controllers/agent/VerifyEmailControllerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+        val request = testSendEmailRequest.withSession(SessionKeys.emailKey -> testEmail)
+        val result = TestVerifyEmailController.sendVerification(request)
+
+        status(result) shouldBe INTERNAL_SERVER_ERROR
+      }
+    }
+
+    "there isn't an email in session" should {
+      "return OK" in {
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply test/controllers/agent/VerifyEmailControllerSpec.scala]: This needs to be addressed

File: app/controllers/agent/ConfirmEmailController.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    extractSessionEmail(agent) match {
+      case Some(email) =>
+        vatSubscriptionService.updateEmail(agent.arn, email) match {
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply app/controllers/agent/ConfirmEmailController.scala]: This needs to be addressed

File: test/controllers/agent/VerifyEmailControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        val request = testSendEmailRequest.withSession(SessionKeys.emailKey -> testEmail)
+        val result = TestVerifyEmailController.sendVerification(request)
+
+        status(result) shouldBe INTERNAL_SERVER_ERROR
+      }
+    }
+
+    "there is not an email in session and the email request returned an unexpected error" should {
+
+      "show the email confirmation page" in {
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply test/controllers/agent/VerifyEmailControllerSpec.scala]: This needs to be addressed

File: test/controllers/agent/VerifyEmailControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+        status(result) shouldBe SEE_OTHER
+        redirectLocation(result) shouldBe Some(routes.ConfirmEmailController.updateEmailAddress().url)
+      }
+    }
+
+
+    "there is an email in session and the email request returned an unexpected error" should {
+
+      "show the email confirmation page" in {
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply test/controllers/agent/VerifyEmailControllerSpec.scala]: This needs to be addressed

File: test/controllers/agent/VerifyEmailControllerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        val result = TestVerifyEmailController.show(request)
+
+        status(result) shouldBe Status.OK
+        contentType(result) shouldBe Some("text/html")
+        charset(result) shouldBe Some("utf-8")
+      }
+    }
+
+    "there isn't an email in session" should {
+      "return OK" in {
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply test/controllers/agent/VerifyEmailControllerSpec.scala]: This needs to be addressed

File: test/controllers/agent/VerifyEmailControllerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    mockAgentOnlyAuthPredicate,
+    messagesApi,
+    mockEmailVerificationService,
+    mockErrorHandler,
+    mockConfig
+  )
+
+  val testVatNumber: String = "999999999"
+  val testEmail: String = "test@email.co.uk"
+  val testContinueUrl: String = "/someReturnUrl/verified"
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply test/controllers/agent/VerifyEmailControllerSpec.scala]: This needs to be addressed

File: test/controllers/agent/VerifyEmailControllerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+  object TestVerifyEmailController extends VerifyEmailController(
+    mockAgentOnlyAuthPredicate,
+    messagesApi,
+    mockEmailVerificationService,
+    mockErrorHandler,
+    mockConfig
+  )
+
+  val testVatNumber: String = "999999999"
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply test/controllers/agent/VerifyEmailControllerSpec.scala]: This needs to be addressed

File: test/controllers/agent/ConfirmEmailControllerSpec.scala [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+  object TestConfirmEmailController extends ConfirmEmailController(
+    mockAgentOnlyAuthPredicate,
+    messagesApi,
+    mockVatSubscriptionService,
+    mockErrorHandler,
+    mockConfig
+  )
+
+  val testVatNumber: String = "999999999"
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply test/controllers/agent/ConfirmEmailControllerSpec.scala]: This needs to be addressed

File: test/views/agent/VerifyEmailViewSpec.scala [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+  lazy implicit val testGetRequest: FakeRequest[AnyContentAsEmpty.type] = FakeRequest("GET", "")
+  val testEmail: String = "test@email.com"
+  lazy val view = views.html.agent.verify_email(testEmail)
+
+  lazy implicit val document: Document = Jsoup.parse(view.body)
+
+  "The Verify Email view" should {
+
+    "have the correct heading" in {
+      document.getElementsByClass("heading-large").text() shouldBe "Verify your email address"
```
[the remark this answers is not in the document]
Comment (TomDudley92) [reply test/views/agent/VerifyEmailViewSpec.scala]: This needs to be addressed

File: it/pages/agent/ConfirmEmailPageSpec.scala [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+      "the Agent is signed up for HMRC-AS-AGENT" when {
+
+        "there is a verified notification email in session" should {
+
+          "Render the Select client VRN view" in {
+
+            val path = "/is-email-verified"
+
+            def isEmailVerified(): WSResponse = get(path, formatNotificationsEmail(Some(notificationsEmail)) ++
```
Comment (TomDudley92) [inline it/pages/agent/ConfirmEmailPageSpec.scala]: Can you remove the parenthesis from this def please. They aren't needed

File: it/pages/agent/ConfirmEmailPageSpec.scala [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+          }
+        }
+
+        "there is a non-verified notification email in session" should {
+
+          "Render the Verify Email view" in {
+
+            val path = "/is-email-verified"
+
+            def isEmailVerified(): WSResponse = get(path, formatNotificationsEmail(Some(notificationsEmail)) ++
```
Comment (TomDudley92) [inline it/pages/agent/ConfirmEmailPageSpec.scala]: Can you remove the parenthesis from this def please. They aren't needed

File: it/pages/agent/ConfirmEmailPageSpec.scala [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+              redirectURI(controllers.agent.routes.SelectClientVrnController.show().url)
+            )
+          }
+        }
+
+        "there is a non-verified notification email in session" should {
+
+          "Render the Verify Email view" in {
+
+            val path = "/is-email-verified"
```
Comment (TomDudley92) [inline it/pages/agent/ConfirmEmailPageSpec.scala]: Could this not be another global val but just named differently? Save having it redefined in all the tests below?

File: it/pages/agent/ConfirmEmailPageSpec.scala [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+            And("I stub an email not verified response from EmailVerificationService")
+            EmailVerificationStub.stubEmailNotVerified
+
+            When("I call the isEmailVerified controller action with a non-verified notification email in session")
+
+            val res = isEmailVerified()
+
+            res should have(
+              httpStatus(SEE_OTHER),
+              redirectURI(controllers.agent.routes.VerifyEmailController.sendVerification().url)
```
Comment (TomDudley92) [inline it/pages/agent/ConfirmEmailPageSpec.scala]: What do you think about having a title / header check here as well? Just to make sure that the correct view is loaded?

File: it/pages/agent/ConfirmEmailPageSpec.scala [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+            And("I stub an email verified successful response from EmailVerificationService")
+            EmailVerificationStub.stubEmailVerified
+
+            When("I call the isEmailVerified controller action with a verified notification email in session")
+
+            val res = isEmailVerified()
+
+            res should have(
+              httpStatus(SEE_OTHER),
+              redirectURI(controllers.agent.routes.SelectClientVrnController.show().url)
```
Comment (TomDudley92) [inline it/pages/agent/ConfirmEmailPageSpec.scala]: What do you think about having a title / header check here as well? Just to make sure that the correct view is loaded?

Comment (davidmthornton) [reply it/pages/agent/ConfirmEmailPageSpec.scala]: Is that possible on a redirect? I did try but couldn't find any html body.

Comment (TomDudley92) [reply it/pages/agent/ConfirmEmailPageSpec.scala]: Good point actually. Ignore it then

**Pull request merged**